i18n-helpers/src/lib.rs

```diff
@@ -12,112 +12,487 @@
+    message_open: bool,
+}
+
+impl<'a> MsgAccumulator<'a> {
+    fn new(document: &'a str) -> Self {
+        Self {
+            document: document,
+            offsets: document
+                .match_indices("\n")
+                .map(|(offset, _)| offset)
+                .collect(),
+            msgs: vec![],
+            depth: 0,
+            skip_until_depth: None,
+            message_open: false,
+        }
+    }
+
+    /// Mark the current message as finished.
+    fn finish_message(&mut self) {
+        self.message_open = false;
+    }
+
+    /// Add a new text message, or extend an existing one.
+    fn push_message(&mut self, span: Range<usize>) {
+        // try to combine with an existing message.
+        if self.message_open {
+            if let Some(last) = self.msgs.last_mut() {
+                last.extend(span.end);
+                return;
+            }
+        }
+
+        self.msgs
+            .push(Message::new(self.line_number(span.start), span));
+        self.message_open = true;
+    }
+
+    /// Calculate the line number for the given offset.
+    fn line_number(&self, offset: usize) -> usize {
+        self.offsets.partition_point(|&o| o < offset) + 1
+    }
+
+    /// Push a new Markdown event into the accumulator.
+    fn push_event(&mut self, evt: Event<'a>, span: Range<usize>) {
+        #[cfg(test)]
+        println!("{evt:?} -- {:?}", &self.document[span.start..span.end]);
+
+        // Track the nesting depth.
+        match evt {
+            Event::Start(_) => self.depth += 1,
+            Event::End(_) => self.depth -= 1,
+            _ => {}
         }
-        None => {
-            if last < trimmed.len() {
-                let result = (lineno, trimmed[last..].trim_end_matches('\n'));
-                last = trimmed.len();
-                Some(result)
-            } else {
-                None
+
+        // Handle skip_until_depth, including skipping the End event that
+        // returned to the desired level.
+        if let Some(depth) = self.skip_until_depth {
+            if self.depth <= depth {
+                self.skip_until_depth = None;
+            }
+            return;
+        }
+
+        match evt {
+            // Consider "inline" tags to be just part of the text.
+            Event::Start(
+                Tag::Emphasis | Tag::Strong | Tag::Strikethrough | Tag::Link(..),
+            ) => self.push_message(span),
+            Event::End(
+                Tag::Emphasis | Tag::Strong | Tag::Strikethrough | Tag::Link(..),
+            ) => self.push_message(span),
+
+            // We want to translate everything: text, code (from backticks, `..`), or HTML.
+            Event::Text(_) | Event::Code(_) | Event::Html(_) => self.push_message(span),
+
+            // For many event types we just take the entire text from Start to End, which is
+            // already encompassed in the event span.
+            Event::Start(Tag::CodeBlock(_))
```

Comment:
You could consider moving all the tags into the `Event::Start` pattern (I don't know it looks nicer): `Event::Start(Tag::CodeBlock(_) | Tag::Heading(..) | ...)`.